core/src/main/scala/akka/kafka/KafkaConsumerActor.scala

```diff
@@ -144,31 +140,43 @@ private[kafka] class KafkaConsumerActor[K, V](settings: ConsumerSettings[K, V])
     super.preStart()
     requests = Map.empty[TopicPartition, ActorRef]
     consumer = settings.createKafkaConsumer()
-    nextScheduledPoll = None
     commitsInProgress = 0
-    pollExpected = false
-    schedulePoll()
   }
 
   override def postStop(): Unit = {
-    nextScheduledPoll.foreach(_.cancel())
+    pollTask.cancel()
     consumer.close()
     super.postStop()
   }
 
   def poll() = {
-    if (pollExpected) {
-      pollExpected = false
-      nextScheduledPoll.foreach(_.cancel())
-      nextScheduledPoll = None
-
-      //set partitions to fetch
-      val partitionsToFetch = requests.keys.toSet
-      consumer.assignment().asScala.foreach { tp =>
-        if (partitionsToFetch.contains(tp)) consumer.resume(java.util.Collections.singleton(tp))
-        else consumer.pause(java.util.Collections.singleton(tp))
+    //set partitions to fetch
+    val partitionsToFetch = requests.keys.toSet
+    consumer.assignment().asScala.foreach { tp =>
+      if (partitionsToFetch.contains(tp)) consumer.resume(java.util.Collections.singleton(tp))
+      else consumer.pause(java.util.Collections.singleton(tp))
+    }
+
+    if (requests.isEmpty) {
+
+      def checkNoResult(rawResult: ConsumerRecords[K, V]): Unit =
+        if (!rawResult.isEmpty)
+          throw new IllegalStateException(s"Got ${rawResult.count} unexpected messages")
+
+      checkNoResult(consumer.poll(0))
+
+      var i = 10
+      while (i > 0 && commitsInProgress > 0) {
+        LockSupport.parkNanos(10 * 1000)
+        val pollTime = System.nanoTime()
+        val pollTimeout = if (i == 1) 1L else 0L
```

Comment:
This implementation does:
* 10 attempts with poll(0)
* 1 attempt with poll(1)
* then scheduled message after 50 ms

We are of course somewhat guessing here. Might be different on other machines.